static-alloc/src/unsync/node.rs

```diff
@@ -0,0 +1,206 @@
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct AllocFailure {
+    /// The size that could not be allocated
+    size: usize,
+}
+
+impl AllocFailure {
+    fn new(size: usize) -> Self {
+        Self { size }
+    }
+}
+
+/// An `AllocationError` represents a failure during the process
+/// of allocating a Node.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub enum AllocationError {
+    /// A failed allocation of a Node
+    AllocationFailure(AllocFailure),
+
+    /// Anything else that went wrong during the proces
+    Other,
+}
+
+impl Node {
+    /// Given `layout`, allocates a Node and returns a pointer.
+    /// The Node will be initialized with zeroes.
+    unsafe fn alloc_raw(layout: Layout) -> Result<*mut u8, AllocFailure> {
+        let ptr = alloc_zeroed(layout);
+
+        if ptr.is_null() {
+            return Err(AllocFailure::new(layout.size()));
+        } else {
+            Ok(ptr)
+        }
+    }
+
+    /// Deallocates `this`
+    pub(crate) unsafe fn dealloc(this: NonNull<Self>) {
+        let size = this.as_ref().capacity();
+        let layout = Self::layout_from_size(size).unwrap();
+
+        dealloc(this.as_ptr() as *mut u8, layout);
+    }
+}
+
+impl Node {
+    /// Returns capacity of this `Node`.
+    /// This is how many *bytes* can be allocated
+    /// within this node.
+    pub(crate) fn capacity(&self) -> usize {
+        self.data.len()
+    }
+
+    pub(crate) fn current_index(&self) -> usize {
+        self.index.get()
+    }
+
+    pub(crate) fn set_next(&self, next: Link) {
+        self.next.set(next);
+    }
+
+    pub(crate) fn take_next(&self) -> Link {
+        self.next.take()
+    }
+
+    /// Allocates a Node and returns it.
+    pub(crate) fn alloc(size: usize) -> Result<NonNull<Self>, AllocationError> {
+        let layout = Self::layout_from_size(size).map_err(|_| AllocationError::Other)?;
+
+        unsafe {
+            let ptr = Self::alloc_raw(layout).map_err(|error| AllocationError::AllocationFailure(error))?;
+
+            let raw_mut: *mut [Cell<MaybeUninit<u8>>] =
+                ptr::slice_from_raw_parts_mut(ptr.cast(), size);
+            let node_ptr = raw_mut as *mut Node;
+
+            Ok(NonNull::new_unchecked(node_ptr))
+        }
+    }
+}
+
+fn next_power_of(n: usize, pow: usize) -> usize {
+    let remain = n % pow;
+
+    [n, n + (pow - remain)][(remain != 0) as usize]
+}
+
+impl Node {
+    /// Returns the start *address* of the data field
+    fn data_start_address(&self) -> usize {
+        self.data.as_ptr() as usize + self.index.get()
+    }
+
+    fn align_index_for<T>(&self) -> usize {
+        let start_addr = self.data_start_address();
+        let aligned_start = next_power_of(start_addr, mem::align_of::<T>());
+        let aligned_index = aligned_start - self.data.as_ptr() as usize;
+        aligned_index
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
```

Comment:
`derive(Ord)` seems like a potential forward compatibility hazard, the other seem fine.